File: Problems/1749. Maximum Absolute Sum of Any Subarray/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxAbsoluteSum(vector<int>& nums) {
        int len = nums.size();
        if (len == 1) return abs(nums[0]);

        int currMax = nums[0], currMin = nums[0];
        int maxSum = nums[0], minSum = nums[0];

        for (int i = 1; i < len; i++) {
            currMax = max(currMax + nums[i], nums[i]);
            maxSum = max(maxSum, currMax);

            currMin = min(currMin + nums[i], nums[i]);
            minSum = min(minSum, currMin);
        }

        return max(abs(maxSum), abs(minSum));
    }
};

File: Problems/1749. Maximum Absolute Sum of Any Subarray/Solution.rs
impl Solution {
    pub fn max_absolute_sum(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        if len == 1 {
            return i32::abs(nums[0]);
        }

        let (mut curr_max, mut curr_min) = (nums[0], nums[0]);
        let (mut max_sum, mut min_sum) = (nums[0], nums[0]);

        for i in 1..len {
            curr_max = nums[i].max(curr_max + nums[i]);
            max_sum = max_sum.max(curr_max);

            curr_min = nums[i].min(curr_min + nums[i]);
            min_sum = min_sum.min(curr_min);
        }

        return i32::abs(max_sum).max(i32::abs(min_sum));
    }
}